py/compile.c

```diff
@@ -800,6 +800,9 @@ STATIC qstr compile_classdef_helper(compiler_t *comp, mp_parse_node_struct_t *pn
     scope_t *cscope = (scope_t*)pns->nodes[3];
 
     // compile the class
+#if MICROPY_PY_SYS_SETTRACE
+    EMIT_ARG(set_source_line, pns->source_line);
```

Comment:
It might make sense to enable these 4 set-source-line emits unconditionally.  I assume they are there to get better line numbering for these cases, is that correct?  Wouldn't such number be useful regardless of settrace being enabled?

Reply:
> It might make sense to enable these 4 set-source-line emits unconditionally.
> Wouldn't such number be useful regardless of settrace being enabled?

Nobody was missing them so far. Unconditional emitters would make the _release_ frozen uPy images larger with _useless information_. On Unix port this doesn't matter but for firmware images I guess it does. I didn't make the measurements though. And I'm not aware of any usefulness outside of `sys.settrace`.

> I assume they are there to get better line numbering for these cases, is that correct? 

Yes. The emitters were added in order to get precision line numbering at places where it was missing -- compared to CPython `sys.settrace` output.

Reply:
> Nobody was missing them so far.

True!



> Unconditional emitters would make the _release_ frozen uPy images larger with _useless information_.

It's possible to compile bytecode with optimisation which doesn't store the line number table.

For this particular line here I couldn't see any difference in generated bytecode with it included.  Do you have a test case where it makes a difference?

Reply:
> For this particular line here I couldn't see any difference in generated bytecode with it included. Do you have a test case where it makes a difference? 

Unfortunately I'm not able to reproduce the testcase for this line. It will be easier to remove it for now and wait for the issue to pop up.

Reply:
Sounds good.